tests/mem_tracking.rs

```diff
@@ -0,0 +1,153 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use core::fmt::Debug;
+use parity_scale_codec::{
+	alloc::{
+		collections::{BTreeMap, BTreeSet, LinkedList, VecDeque},
+		rc::Rc,
+	},
+	DecodeWithMemLimit, DecodeWithMemTracking, Encode, Error, MemTrackingInput,
+};
+use parity_scale_codec_derive::{
+	Decode as DeriveDecode, DecodeWithMemTracking as DeriveDecodeWithMemTracking,
+	Encode as DeriveEncode,
+};
+
+const ARRAY: [u8; 1000] = [11; 1000];
+
+#[derive(DeriveEncode, DeriveDecode, DeriveDecodeWithMemTracking, PartialEq, Debug)]
+#[allow(clippy::large_enum_variant)]
+enum TestEnum {
+	Empty,
+	Array([u8; 1000]),
+}
+
+#[derive(DeriveEncode, DeriveDecode, DeriveDecodeWithMemTracking, PartialEq, Debug)]
+struct ComplexStruct {
+	test_enum: TestEnum,
+	boxed_test_enum: Box<TestEnum>,
+	box_field: Box<u32>,
+	vec: Vec<u8>,
+}
+
+fn decode_object<T>(obj: T, mem_limit: usize, expected_used_mem: usize) -> Result<T, Error>
+where
+	T: Encode + DecodeWithMemTracking + DecodeWithMemLimit + PartialEq + Debug,
+{
+	let encoded_bytes = obj.encode();
+
+	let decoded_obj = T::decode_with_mem_limit(&mut &encoded_bytes[..], mem_limit)?;
+	assert_eq!(&decoded_obj, &obj);
+
+	let raw_input = &mut &encoded_bytes[..];
+	let mut input = MemTrackingInput::new(raw_input, mem_limit);
+	let decoded_obj = T::decode(&mut input)?;
+	assert_eq!(&decoded_obj, &obj);
+	assert_eq!(input.used_mem(), expected_used_mem);
+	if expected_used_mem > 0 {
+		let raw_input = &mut &encoded_bytes[..];
+		let mut input = MemTrackingInput::new(raw_input, expected_used_mem);
```

Comment:
Why does it fail with this? I thought that it should fail only with `expected_used_mem - 1`.